Arcyd: Add reason for stopping to info logs

cs-shadow opened the pull request:
When Arcyd stops gracefully, a log entry is left in /var/log/info. This
is useful for knowing when Arcyd was stopped. It would be helpful to
have a reason in the log as well. This commit adds command line
argument "-m"/"--message" to "arcyd stop" and "arcyd restart" commands
to supply this reason.

This argument is optional but it is highly recommended to provide this
argument whenever stopping or restarting Arcyd.

Test plan:
$ ./precommit.sh

Launch Arcyd test shell using ./testbed/arcyd/test_shell.sh. Now, stop
Arcyd using following command:
  $ arcyd stop -m "Stopped arcyd to ensure this message is logged"

Now ensure that /var/log/info contains an entry similar to the
following:
  INFO: Reason for stopping Arcyd: Stopped arcyd to ensure this message is logged
  INFO: stopped arcyd

Start Arcyd again using following command:
  $ arcyd start

Now, restart Arcyd using following command:
  $ arcyd restart -m "Restarted arcyd to ensure this message is logged"

Now ensure that /var/log/info contains an entry similar to the
following:
  INFO: Reason for stopping Arcyd: Restarted arcyd to ensure this message is logged
  INFO: stopped arcyd
  INFO: started arcyd

Fixes #39

aevri commented:
After reading the patch, I now realise I should have recommended a different approach to logging the reason. Sorry :)

Although we wait until after Arcyd has stopped to put the reason for stopping in the logs, we are still potentially racing against another contender to write to the log files.

Also, if the user `ctrl-c`'s after the kill file is written then the reason will not be written to the logs.

An alternate approach we could take is to write the reason into the kill file, and have the running arcyd instance do the logging instead. That should address the above concerns.

Please could you address those concerns by writing into the kill file or another approach?

cs-shadow commented:
I've updated when the reason for stopping is logged. Also added a commit which slightly refactors the while loop in `do()` by removing the usage of `finished` variable.

Note: With these set of changes, the actual message that arcyd is stopped and the reason for stopping are logged from different places and also the reason comes before the message that arcyd is stopped. Let me know if you think it should be the other way round. For me personally, both ways are acceptable.

cs-shadow commented:
As per our previous discussions, we need to write to killfile atomically and that is being added in #52. Will update this PR once that change gets in.

cs-shadow commented:
This is ready for review now but `abdi_processrepoarglist:do()` is still bordering on 20 local vars so changes are not very elegant. Any suggestions to fix that are welcome.

aevri commented on py/abd/abdcmd_stop.py:
```diff
@@ -25,11 +25,14 @@ def getFromfilePrefixChars():
+    parser.add_argument(
+        '-m',
+        '--message',
+        help="reason for stopping arcyd")
```
The default value for `default` in `add_argument()` is `None`, which `phlsys_fs.write_text_file_atomic()` doesn't like as a parameter :)

aevri commented on py/abd/abdcmd_restart.py:
```diff
@@ -37,13 +37,18 @@ def setupParser(parser):
+    parser.add_argument(
+        '-m',
+        '--message',
+        help="reason for restarting arcyd")
```
we need `default=''` for this one too

cs-shadow replied:
Fixed now

aevri commented:
Great, thanks!
